Public/Src/Tools/VBCSCompilerLogger/VBCSCompilerLogger.cs

```diff
@@ -0,0 +1,191 @@
+
+            // If the output filename is not specified, we follow the logic documented for csc.exe
+            // https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/out-compiler-option
+            // If you do not specify the name of the output file:
+            // - An.exe will take its name from the source code file that contains the Main method.
+            // - A.dll or .netmodule will take its name from the first source code file.
+            // Note: observe however that when invoked through the standard MSBuild csc task, the output file name is always specified
+            // and it is based on the first source file (when the original task didn't specify it). So this is just about being conservative.
+            if (string.IsNullOrEmpty(outputFileName))
+            {
+                switch (args.CompilationOptions.OutputKind)
+                {
+                    // For these cases, the first source file is used
+                    case OutputKind.DynamicallyLinkedLibrary:
+                        outputFileName = Path.ChangeExtension(args.SourceFiles[0].Path, ".dll");
+                        break;
+                    case OutputKind.NetModule:
+                        outputFileName = Path.ChangeExtension(args.SourceFiles[0].Path, ".netmodule");
+                        break;
+                    case OutputKind.WindowsRuntimeMetadata:
+                        outputFileName = Path.ChangeExtension(args.SourceFiles[0].Path, ".winmdobj");
+                        break;
+                    // For these cases an .exe will be generated based on the source file that contains a Main method.
+                    // We cannot easily predict this statically, so we bail out for this case.
+                    case OutputKind.ConsoleApplication:
+                    case OutputKind.WindowsApplication:
+                    case OutputKind.WindowsRuntimeApplication:
+                        throw new InvalidOperationException("The output filename was not specified and it could not be statically predicted. Static predictions are required for managed compilers when shared compilation is enabled. " +
+                            "Please specify the output filename or disable shared compilation by setting 'useManagedSharedCompilation' in Bxl main configuration file.");
+                    default:
+                        throw new InvalidOperationException($"Unrecognized OutputKind: ${args.CompilationOptions.OutputKind}");
+                }
+            }
+
+            Contract.Assert(!string.IsNullOrEmpty(outputFileName));
+            return outputFileName;
+        }
+
+        private void RegisterOutput(string filePath)
+        {
+            RegisterAccess(filePath, m_augmentedReported.TryReportFileCreations);
+        }
+
+        private void RegisterInput(string filePath)
+        {
+            RegisterAccess(filePath, m_augmentedReported.TryReportFileReads);

```

Comment:
No idea on the perf / asyncness of TryReportFileReads. Should inputs be batched rather than sent one by one? Or it doesn't really matter?

Reply:
It shouldn't matter. This writes into an already opened pipe. There is in fact logic on bxl side (the other end of the pipe) to do async reads

---
In reply to: [346069206](https://github.com/microsoft/BuildXL/pull/1207#discussion_r346069206) [](ancestors = 346069206)